File: trunk/Database/nastya/src/database/index/DatabaseKey.java
package database.index;

import tree.Key;
import tree.TreeElement;
import dbentities.Condition;

/**
 * Ключ дерева, используемый при работе с текущей базой данных
 *
 * @author nastya
 * Date: 28.08.2009
 * Time: 14:02:10
 *
 */
public class DatabaseKey extends Key{

    Condition condition;

    public DatabaseKey(Condition condition, TreeElement link) {
        super(link);
        this.condition = condition;
    }

    public DatabaseKey(Condition condition) {
        this(condition, null);
    }


    public Key clone() {
        return new DatabaseKey(condition.clone(), link);
    }

    public int compareTo(Object o) {
        if(o == null) return 1;
        if(o.getClass() != getClass()) return 1;
        DatabaseKey k = (DatabaseKey)o;

        return condition == null ? -1 : condition.compareTo(k.condition);
    }

    @Override
    public String toString() {
        return "DatabaseKey(" + condition + ")";
    }
}

File: branches/2008-2009/Database/Babaka/DB/DB3/Index/DatabaseKey.cs
using System;
using DB3.Entities;
using DB3.Interfaces;
using DB3.Tree;

namespace DB3.Index
{
    [Serializable]
    class DatabaseKey : Key
    {
        public Condition condition
        {
            get; private set;
        }

        public DatabaseKey(Condition condition, ITree link)
            : base(link)
        {
            this.condition = condition;
        }
        public DatabaseKey(Condition condition)
            : base(null)
        {
            this.condition = condition;
        }

        public override object Clone()
        {
            return new DatabaseKey(condition.Clone() as Condition, Link);
        }
        public override int CompareTo(object o)
        {
            if (o == null)
                return 1;
            if (o.GetType() != GetType())
                return 1;
            DatabaseKey dbk = (DatabaseKey)o;

            return condition == null ? -1 : condition.CompareTo(dbk.condition);
        }

        //public override bool Equals(object o)
        //{
        //    return CompareTo(o) == 0;
        //}

        //public static bool operator ==(DatabaseKey fst, DatabaseKey snd)
        //{
        //    try
        //    {
        //        return fst.Equals(snd);
        //    }
        //    catch (NullReferenceException)
        //    {
        //        try
        //        {
        //            if (snd.CompareTo(null) != 0)
        //            {
        //                return false;
        //            }
        //        }
        //        catch (Exception)
        //        {
        //            return true; 
        //        }
        //        return true; 
        //    }           
        //}

        //public static bool operator !=(DatabaseKey fst, DatabaseKey snd)
        //{
        //    return !(fst == snd);
        //}

        //public static bool operator >(DatabaseKey fst, DatabaseKey snd)
        //{
        //    return fst.condition > snd.condition;
        //}

        //public static bool operator <(DatabaseKey fst, DatabaseKey snd)
        //{
        //    return fst.condition < snd.condition;
        //}

        //public override string ToString()
        //{
        //    return "DatabaseKey(" + condition + ")";
        //}

        public bool Equals(DatabaseKey other)
        {
            if (ReferenceEquals(null, other)) return false;
            if (ReferenceEquals(this, other)) return true;
            return base.Equals(other) && Equals(other.condition, condition);
        }

        public override int GetHashCode()
        {
            return base.GetHashCode() + condition.GetHashCode();
        }
    }
}
